### DISCUSSION TITLE
Geometric MTFs

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#30] [General]
Is this available or planned?

**Comment by mjhoptics** [comment]
MTF calculations aren't available now. I suspect I'll add it eventually, but I don't have any test examples with validated results that I can compare against so that makes it more challenging.

**Comment by dibyendumajumdar** [comment]
hi @mjhoptics 

I recently tried implementing geometric MTF based on inputs from ChatGPT(!) - I had surprisingly good results. So I wanted to share the steps here:

* Compute x/y ray intercepts from a spot diagram. I am using hexapolar pattern and adjusting to centroid.
* Compute a 2d histogram. I am using 512 bins; and pixel_size 0.001.
* Normalize histogram
* Compute line spread functions for x and y by summing up values by axis in the histogram
* Normalize each LSF
* Compute complex forward FFT on LSF using LSF as the real values
* Compute the FFT magnitude (for only half the FFT array) and normalize  - this then is the output
* Compute the frequencies corresponding to each value in the magnitude

Above gives the Geo MTF per field per wavelength.

Then to compute polychromatic MTF:

* Use the FFTs calculated above for the wavelengths for a field and do a weighted sum for each real/imaginary value - weight by wavelength weight - this gives a new FFT
* Then compute magnitude in the same way as before.

Attached is a writeup from ChatGPT
[Replicating Zemax Geometric MTF.docx](https://github.com/user-attachments/files/24295798/Replicating.Zemax.Geometric.MTF.docx)

**Reply by Meto423** [reply]
Hello, do you have a chance to share the source code of the MTF calculation. I've been struggling with it for some time. 
Thanks.

**Reply by dibyendumajumdar** [reply]
Hi,

My implementation is in my Java port of ray-optics. You can find it [here](https://github.com/BeamFour/rayoptics4j/tree/main/src/main/java/org/redukti/rayoptics/analysis).

It would be nice to have a Python version - its on my todo list but haven't got around to it. 

If you show me your version and describe what issue you are facing I can try to help, but my Python is not great.

**Comment by dibyendumajumdar** [comment]
Couple of additional changes I made since above:

* When adjusting ray intercepts to centroid, compute centroid of reference wavelength and apply to all - rather than separately for each wavelength. Apparently this is important for polychromatic mtf.
* Pad the LSF to twice the size with zeros at both ends - this is what Zemax does apparently. It is to do with ensuring correct FFT.

**Comment by mjhoptics** [comment]
There are 2 Jupyter notebooks working through the calculation of both geometric and diffraction-based MTFs. These were developed last year, when addressing issue #182.
[Diffraction MTF](https://github.com/mjhoptics/ray-optics-notebooks/blob/master/issues/issue182_working_fnum/Diffraction%20MTF.ipynb)
[Geometrical MTF](https://github.com/mjhoptics/ray-optics-notebooks/blob/master/issues/issue182_working_fnum/Geometrical%20MTF.ipynb)
I haven't included them in ray-optics yet largely because I don't have any test results to compare against.
Mike Hayford

**Reply by dibyendumajumdar** [reply]
I had looked at above when I was implementing - I think the Geometric MTF uses approximation described in Modern Optical Engineering? 

According to ChatGPT  Zemax OpticStudio uses FFT for geometric MTF - not the approximation. I haven't tried the approximation approach - I had good results with the ChatGPT suggested approach.

**Comment by mjhoptics** [comment]
I believe they're effectively the same thing. The 1st edition of MOE was
published around the same time as the FFT paper.

**Reply by dibyendumajumdar** [reply]
I think Optiland also uses the approximation approach for Geometric MTF.
For validating - I guess you could compare with their implementation?

**Comment by mjhoptics** [comment]
I think you'll find the FFT is an approximation also. Both approaches are
based on the same math, the implementation is packaged differently.

**Reply by dibyendumajumdar** [reply]
I see, good to know!

What I also do in my implementation - is generate MTFs for 10 fields across the image, average the MTF values across wavelengths for those fields for selected frequencies (10,30,50) and then plot them. This gives results such as published by various lens manufacturers.
